Run workflows depending on changed files

4e6 opened the pull request:
### Pull Request Description

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

close #11579 

Changelog:
- update: add `changed-files` prerequisites to the existing workflows to dispatch them conditionally on files changed when PR is opened and unconditionally on push to `develop`:
  + `gui-pull-request.yml` GUI checks
  + `wasm-pull-request.yml` WASM checks
  + `engine-pull-request.yml` Engine checks
  + `gui-packaging-pull-request.yml` GUI packaging 
 - update: mark `macOS` jobs with `continue-on-error: true` to make them optional (currently they are not required to merge to PR). Even if they fail, the final report will be successful. 
 - refactor: rename workflow files to match their contents:
   + `wasm-checks.yml` WASM Checks (previously `gui-tests.yml` GUI Check)
   + `engine-checks.yml` Engine Checks (previously `scala-new.yml` Engine CI)
   + `gui-packaging.yml` GUI Packaging (previously `gui.yml` GUI Packaging)

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

This PR introduces multiple `*-pull-request.yml` workflows which are dispatched on pull requests and pushes to the `develop` branch, calling all other existing workflows (`*-checks.yml`) based on the files changed.

The final `Required Checks` step is required to gather the results of all (possibly skipped) checks. This will be the only (together with Cangelog and the Code Formatting) required check to merge the PR.

```
+---------------------------------------------------------------------------------------------------+
|   GUI                                                                                             |
|   gui-pull-request.yml                 +----------------------------+                             |
|                                   +--->| Storybook                  |---+                         |
|                                   |    | storybook.yml              |   |   +-----------------+   |
|                                   |    +----------------------------+   +-->| Required Checks |   |
|   +--------------------------+    |    +----------------------------+   |   +-----------------+   |
|   | GUI Files Changed        +----+--->| GUI Checks                 +---+                         |
|   | gui-changed-files.yml    |         | gui-checks.yml             |                             |
|   +--------------------------+         +----------------------------+                             |
|                                                                                                   |
+---------------------------------------------------------------------------------------------------+
                                                                                                     
                                                                                                     
+---------------------------------------------------------------------------------------------------+
|   WASM                                                                                            |
|   wasm-pull-request.yml                                                                           |
|                                                                                                   |
|   +--------------------------+         +----------------------------+       +-----------------+   |
|   | WASM Files Changed       +-------->+ WASM Checks                +------>| Required Checks |   |
|   | wasm-changed-files.yml   |         | wasm-checks.yml            |       +-----------------+   |
|   +--------------------------+         +----------------------------+                             |
|                                                                                                   |
+---------------------------------------------------------------------------------------------------+
                                                                                                     
                                                                                                     
+---------------------------------------------------------------------------------------------------+
|   Engine                                                                                          |
|   engine-pull-request.yml                                                                         |
|                                                                                                   |
|   +--------------------------+         +----------------------------+       +-----------------+   |
|   | Engine Files Changed     +----+--->| Engine Checks              +------>| Required Checks |   |
|   | engine-changed-files.yml |    |    | engine-checks.yml          |       +-----------------+   |
|   +--------------------------+    |    +----------------------------+                             |
|                                   |    +----------------------------+                             |
|                                   +--->| Engine Checks (macOS)      |                             |
|                                        | engine-checks-optional.yml |                             |
|                                        +----------------------------+                             |
|                                                                                                   |
+---------------------------------------------------------------------------------------------------+
                                                                                                     
                                                                                                     
+---------------------------------------------------------------------------------------------------+
|   GUI Packaging                                                                                   |
|   gui-packaging-pull-request.yml                                                                  |
|                                                                                                   |
|   +--------------------------+         +----------------------------+       +-----------------+   |
|   | GUI Files Changed        +----+--->| GUI Packaging              +------>| Required Checks |   |
|   | gui-changed-files.yml    |    |    | gui-packaging.yml          |       +-----------------+   |
|   +--------------------------+    |    +----------------------------+                             |
|   +--------------------------+    |    +----------------------------+                             |
|   | Engine Files Changed     +----+--->| GUI Packaging (macOS)      |                             |
|   | engine-changed-files.yml |         | gui-packaging-optional.yml |                             |
|   +--------------------------+         +----------------------------+                             |
|                                                                                                   |
+---------------------------------------------------------------------------------------------------+
```

4e6 committed:
update: CI PR workflows

4e6 committed:
fix: workflow call

4e6 committed:
fix: report-success

4e6 committed:
fix: wasm check permissions

4e6 committed:
fix: concurrency settings

4e6 committed:
update: final report

4e6 committed:
update: single checks report

4e6 committed:
misc: cleanup

4e6 committed:
misc: wasm dependency

JaroslavTulach commented on .github/workflows/gui-pull-request.yml:
```diff
@@ -1,6 +1,6 @@
 # This file is not auto-generated. Feel free to edit it.
 
-name: ✨ GUI Pull Request
+name: ✨ Pull Request
```
I guess you want to rename the filename as well.

4e6 committed:
update: print files changed

4e6 committed:
update: always run on develop

4e6 committed:
update: concurrency

4e6 committed:
update: changed files

4e6 committed:
update: macos continue on error

4e6 committed:
DEBUG: continue-on-error

4e6 committed:
DEBUG: speedup continue-on-error

4e6 committed:
DEBUG: continue-on-error 1

4e6 committed:
DEBUG: revert continue-on-error

4e6 committed:
update: gui macos continue on error

4e6 committed:
refactor: rename pull request workflow

4e6 committed:
refactor: rename workflows

4e6 committed:
update: pull request workflow

4e6 committed:
misc: cleanup debug workflow

4e6 committed:
update: finishing touches

4e6 committed:
misc: naming

4e6 commented:
Repo settings should be updated to reflect the changes. In the branch protection rules the following required checks should be 

removed:
- `Build Backend*`
- `Engine (GraalVM CE)*`
- `GUI build*`
- `JVM Tests (GraalVM CE)*`
- `Lint*`
- `Native Rust tests*`
- `Package New IDE*`
- `Standard Library Tests*`
- `Verify License Packages*`

renamed:
- `GUI Checks Success or Skipped` -> `Required Checks`

4e6 committed:
feat: optional targets

4e6 committed:
misc: naming

4e6 committed:
update: reuse gui-changed-files

4e6 committed:
update: speedup

4e6 committed:
DEBUG: git-changed-files unnamed

4e6 committed:
update: reuse changed files

4e6 committed:
update: split pull requests

4e6 committed:
fix: shell script

4e6 committed:
misc: cleanup

4e6 committed:
misc: naming

4e6 committed:
misc: cleanup shader-tools workflow

4e6 committed:
update: changed files patterns

4e6 committed:
Merge branch 'develop' into wip/db/11579-run-ci-backend-when-sources-changed

4e6 committed:
misc: optimize naming

hubertp reviewed (approved):
#yolo

4e6 committed:
Merge branch 'develop' into wip/db/11579-run-ci-backend-when-sources-changed

4e6 committed:
fix: after merge

4e6 committed:
update: regen workflows

4e6 committed:
update: check prettier outcome
